From c2e40e897d0035ad66c78929184b185fc6babd5f Mon Sep 17 00:00:00 2001
From: Justin Ku <chunku@qti.qualcomm.com>
Date: Tue, 24 Mar 2026 05:45:54 -0700
Subject: [PATCH] [linux net] pon_iftrap_process_tx() requires
 bottom_dev(physical)

bottom_dev: the underlying physical interface
Packets may be dropped or consumed during Qdisc processing
so pon_iftrap_process_tx() should be executed after that.

Change-Id: Iace63bdeca69a0fc16c2ccdb23e2b305a2626f46
Signed-off-by: Justin Ku <chunku@qti.qualcomm.com>
---
 net/core/dev.c | 15 ++++++++-------
 1 file changed, 8 insertions(+), 7 deletions(-)

diff --git a/net/core/dev.c b/net/core/dev.c
index 87a19469f0e0..9741433edb99 100644
--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -4788,13 +4788,6 @@ bool dev_fast_xmit_qdisc(struct sk_buff *skb, struct net_device *top_qdisc_dev,
 	 */
 	rcu_read_lock_bh();
 
-#ifdef CONFIG_IPQ_PON
-	if (pon_iftrap_process_tx(skb, DEV_FAST_XMIT_QDISC)) {
-		rcu_read_unlock_bh();
-		return true;
-	}
-#endif
-
 	txq = netdev_core_pick_tx(top_qdisc_dev, skb, NULL);
 	q = rcu_dereference_bh(txq->qdisc);
 	if (unlikely(!q->enqueue)) {
@@ -4831,6 +4824,14 @@ bool dev_fast_xmit_qdisc(struct sk_buff *skb, struct net_device *top_qdisc_dev,
 	/* Update the dev so that we can transmit to bottom device after qdisc */
 	skb->dev = bottom_dev;
 	skb->fast_qdisc = 1;
+
+#ifdef CONFIG_IPQ_PON
+	if (pon_iftrap_process_tx(skb, DEV_FAST_XMIT_QDISC)) {
+		rcu_read_unlock_bh();
+		return true;
+	}
+#endif
+
 	rc = __dev_xmit_skb_qdisc(skb, q, top_qdisc_dev, txq);
 
 	rcu_read_unlock_bh();
-- 
2.34.1

